lib/charms/opensearch/v0/opensearch_distro.py

```diff
@@ -433,9 +434,10 @@ def current(self) -> Node:
         except OpenSearchHttpError:
             # we try to get the most accurate description of the node
             conf_on_disk = self.config.load("opensearch.yml")
+
             return Node(
                 name=self._charm.unit_name,
-                roles=conf_on_disk["node.roles"],
+                roles=conf_on_disk.get("node.roles") or [],
```

Comment:
Roles cannot be null / empty, as we rely on those elsewhere in the code - this will result in unpredictability in the code. 

These are the sources of getting the roles from, by order of priority and availability:
1. check API (as done in line 423)
2. if not available - check Disk as done in this line 
3. if not available - check `opensearch_peer_cm.deployment_desc().config.roles`
4. if not available - check where roles are queried too early and fix it

Reply:
Sure.

In case NO roles found at all: raising an exception?

Out of curiousity. In the use-case where we experience the [issue](https://github.com/canonical/opensearch-operator/issues/222) (departing unit) where should the role come from?
If it is the config, I would adjust the config in the test space to include roles, simulating closer to reality.

Thanks.

Reply:
Question #1:

> 4. if not available - check where roles are queried too early and fix it

I'm confused how this may apply (i.e. "too early") given that we are getting the error as a node is departing from the cluster. 

As the [issue](https://github.com/canonical/opensearch-operator/issues/222) trace indicates, the current error is triggered in attempt to delete the current node (`opensearch_node_exclusions` -> `delete_current`)

Could you please confirm roles MUST be available until the last moment the unit exists, including the pattern above? I.e. after quitting the `peer` relation, etc. 

(In case not: how should we act then?)

Question #2:

We have 2 other attributes (`node.attr.app_id`, `node.attr.temp`) similarly used in this workflow. Are they supposed to be defined the same way (and sources) as `node.roles`? If yes, we probably want to apply the same checks for all.

Reply:
> Well the current code has already 1 and 2 above. We need to add the 3rd check. If it still fails, it means that the call happens too early - in which case, the call needs to be made later in the code.

That part is clear :-)  My question was targeting the [issue](https://github.com/canonical/opensearch-operator/issues/222) specifically. I try to make it more clear.

In this particular case (i.e. departing node already offline) where do we expect to find the roles? In other words, typically for example do we ALWAYS have roles configured statically (in `opensearch.yaml`)

I would like to adjust test environment accordingly, i.e. providing coverage for the bug fix according to the expected behavior of the product.

Reply:
Summarizing the offline conversation:

Two worklows may be impacted: initialization and shutdown process. 
A node should be receiving `roles` information in the following order from the following sources.

Initialization:
 1. `storage-attached` (i.e. Openserach gets its disk space)
 2. roles information is written to static config
 3. roles information is populated on the `peer` relation describing the `Deployment` configuration
 4. once the service is up, the Opensearch API is used for roles information

Shutdown (2-3 order may be reverse):
 1. Opensearch API may not be available anymore
     - either as the unit may get offline 
     - or there may be no Opensearch online unit to contact anymore
 2. `storage-detached` (no more config file or disk space)
 3. peer relation departed (no more `Deployment` configuration

The bug is raised around 2-3. In addition we should check if before 2) the static config may not get wiped out.